pkg/controller/volume/pvcprotection/pvc_protection_controller.go

```diff
@@ -290,6 +290,10 @@ func (c *Controller) podUsesPVC(logger klog.Logger, pod *v1.Pod, pvc *v1.Persist
 	return false
 }
 
+func podIsFinished(pod *v1.Pod) bool {
+	return pod.Status.Phase == v1.PodFailed || pod.Status.Phase == v1.PodSucceeded
```

Comment:
When `Phase == PodSucceeded || PodFailed`, does it mean that the pod's volumes were already unmounted? Because if not, the PVC is still needed.

Reply:
/sig node
/hold

Reply:
> When `Phase == PodSucceeded || PodFailed`, does it mean that the pod's volumes were already unmounted? Because if not, the PVC is still needed.

Yes, it does.
See at https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go#L254
and https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/volumemanager/cache/actual_state_of_world.go#L975.

Reply:
We cannot make sure that all volumes are unmounted when pod's phase is `Succeeded` or `Failed`. The umount process is asynchronous with the pod's phase change. The kubelet wait for the umount process to finish in the [SyncTerminatedPod](https://github.com/kubernetes/kubernetes/blob/dc3f5ec6ccb9855dfa99f4c1078625df5fdfab6a/pkg/kubelet/kubelet.go#L2146) func. Before the func is called, the pod phase maybe `Succeeded` or `Failed`.

/cc @pacoxu 

Can you help to review this? Thanks.

Reply:
/cc @kannon92 @SergeyKanzhelev 



> When `Phase == PodSucceeded || PodFailed`, does it mean that the pod's volumes were already unmounted? Because if not, the PVC is still needed.

I agree with this comments. If the PVC is not unmounted, we should not just ignore it.

Can we just accelerate the unmount process in kubelet to make the PVC deleting success earlier?

Reply:
> Can we just accelerate the unmount process in kubelet to make the PVC deleting success earlier?

@pacoxu  Do you meaning call `kl.statusManager.SetPodStatus(pod, apiPodStatus)` after `kl.volumeManager.WaitForUnmount(ctx, pod)`?

Reply:
> We cannot ensure that all volumes are unmounted when the pod's phase is `Succeeded` or `Failed`. The umount process is asynchronous with the pod's phase change. The kubelet waits for the umount process to finish in the [SyncTerminatedPod](https://github.com/kubernetes/kubernetes/blob/dc3f5ec6ccb9855dfa99f4c1078625df5fdfab6a/pkg/kubelet/kubelet.go#L2146) func. Before the function is called, the pod phase may be `Succeeded` or `Failed`.

You are right, the pod phase can be marked as `Succeeded` or `Failed` in :https://github.com/kubernetes/kubernetes/blob/dc3f5ec6ccb9855dfa99f4c1078625df5fdfab6a/pkg/kubelet/kubelet.go#L1723.
When volumes unmount, updating a pod condition named `PodTerminated` may solve the problem.

Reply:
> When volumes unmount, updating a pod condition named `PodTerminated` may solve the problem.

I am +1 for this proposal. But, we still needs more input from other sig node owners.
/cc @kannon92 @haircommander

Reply:
I think we may also consider generic ephemeral volumes, which are the volumes that are created and destroyed with the pods. If the pod is terminated, it won't be reclaimed. 
related-to: https://github.com/kubernetes/kubernetes/issues/123592.

Reply:
Is the question "should we add a state named PodTerminated for this case?" if so:  the question would be better asked at SIG-node first and I think we'd need to go through the KEP processes

Reply:
@haircommander I submitted a KEP(https://github.com/kubernetes/enhancements/pull/4577), can you take a look？

Reply:
cc @pacoxu @carlory